pkg/middleware/auth_proxy/auth_proxy.go

```diff
@@ -0,0 +1,320 @@
+	models "github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+const (
+
+	// CachePrefix is a prefix for the cache key
+	CachePrefix = "auth-proxy-sync-ttl:%s"
+)
+
+// AuthProxy struct
+type AuthProxy struct {
+	store  *remotecache.RemoteCache
+	ctx    *models.ReqContext
+	orgID  int64
+	header string
+
+	LDAP func(server *login.LdapServerConf) login.ILdapAuther
+
+	enabled     bool
+	whitelistIP string
+	headerType  string
+	headers     map[string]string
+	cacheTTL    int
+	ldapEnabled bool
+}
+
+// Error auth proxy specific error
+type Error struct {
+	Message      string
+	DetailsError error
+}
+
+// newError creates the Error
+func newError(message string, err error) *Error {
+	return &Error{
+		Message:      message,
+		DetailsError: err,
+	}
+}
+
+// Error returns a Error error string
+func (err *Error) Error() string {
+	return fmt.Sprintf("%s", err.Message)
+}
+
+// Options for the AuthProxy
+type Options struct {
+	Store *remotecache.RemoteCache
+	Ctx   *models.ReqContext
+	OrgID int64
+}
+
+// New instance of the AuthProxy
+func New(options *Options) *AuthProxy {
+	header := options.Ctx.Req.Header.Get(setting.AuthProxyHeaderName)
+
+	return &AuthProxy{
+		store:  options.Store,
+		ctx:    options.Ctx,
+		orgID:  options.OrgID,
+		header: header,
+
+		LDAP: login.NewLdapAuthenticator,
+
+		enabled:     setting.AuthProxyEnabled,
+		headerType:  setting.AuthProxyHeaderProperty,
+		headers:     setting.AuthProxyHeaders,
+		whitelistIP: setting.AuthProxyWhitelist,
+		cacheTTL:    setting.AuthProxyLdapSyncTtl,
+		ldapEnabled: setting.LdapEnabled,
+	}
+}
+
+// IsEnabled checks if the proxy auth is enabled
+func (auth *AuthProxy) IsEnabled() bool {
+
+	// Bail if the setting is not enabled
+	if auth.enabled == false {
+		return false
+	}
+
+	return true
+}
+
+// HasHeader checks if the we have specified header
+func (auth *AuthProxy) HasHeader() bool {
+	if len(auth.header) == 0 {
+		return false
+	}
+
+	return true
+}
+
+// IsAllowedIP compares presented IP with the whitelist one
+func (auth *AuthProxy) IsAllowedIP() (bool, *Error) {
+	ip := auth.ctx.Req.RemoteAddr
+
+	if len(strings.TrimSpace(auth.whitelistIP)) == 0 {
+		return true, nil
+	}
+
+	proxies := strings.Split(auth.whitelistIP, ",")
+	var proxyObjs []*net.IPNet
+	for _, proxy := range proxies {
+		result, err := coerceProxyAddress(proxy)
+		if err != nil {
+			return false, newError("Could not get the network", err)
+		}
+
+		proxyObjs = append(proxyObjs, result)
+	}
+
+	sourceIP, _, _ := net.SplitHostPort(ip)
+	sourceObj := net.ParseIP(sourceIP)
+
+	for _, proxyObj := range proxyObjs {
+		if proxyObj.Contains(sourceObj) {
+			return true, nil
+		}
+	}
+
+	err := fmt.Errorf(
+		"Request for user (%s) from %s is not from the authentication proxy", auth.header,
+		sourceIP,
+	)
+
+	return false, newError("Proxy authentication required", err)
+}
+
+// InCache checks if we have user in cache
+func (auth *AuthProxy) InCache() bool {
```

Comment:
Everytime this func and/or `GetUserIDViaCache()` is called we will hit the remote cache. Would be nice to rewrite this to only hit the remote cache once.